### ISSUE TITLE
Format of Machine Learning API

### ISSUE THREAD

**Issue by Kelvin-Ng** [husky-team/husky#186]
I have thought of several choices, but I don't have a clue of which one is absolutely better. Currently my preference is in descending order. Discussion may help for deciding the final choice.

## 1. Problem-based
Basically, each machine learning problem is an optimization problem trying to optimize a function. Each problem can be made as an object. This object is responsible for telling others about the structure and characteristics of the problem. So, this object is in the form of the following (for example only):

```C++
class XXXProblem {
public:
    double cost_func();
    Vector cost_func_grad();
    Vector cost_func_grad(int i); // if the objective function can be expressed as f(x) = f_1(x) + f_2(x) +...+ f_n(x), it returns gradient of f_i(x)
    double regularization_func(); // seems that it is not needed in most cases
    double regularization_func_grad();
    double prox_func();
    void update_w(Vector updates); // the problem may want to update the weights in some special ways, such as dropout. If gradient descent is used to optimize this problem, the gradient will be passed as updates
private:
    Matrix data; // matrix is just an abstract type. It could be an object list of vector etc.
    Vector w;
};
```

Users may want to only implement a subset of the above APIs, depending on what optimization algorithms they want to use. However, machine learning algorithms inside husky lib should implement all of them.

Then, to optimize a problem, first initialize the Problem object, and then call something like `some_optimization_algo(problem, some, parameters, for, that, algo)`.

Note that the object is an object for an optimization problem, **not** a machine learning model. For logistic regression, for example, L1 norm and L2 norm regularized are two different optimization problems, and so correspond to two different `Problem` classes. However, this can be solved by doing inheritance.

## 2. Model-based
This is indeed complicated. Roughly speaking, we construct such a class, for example:

```C++
class XXXModel {
public:
    double cost_func();
    Vector cost_func_grad();
    Vector cost_func_grad(int i); // if the objective function can be expressed as f(x) = f_1(x) + f_2(x) +...+ f_n(x), it returns gradient of f_i(x)
    void update_w(Vector updates); // the problem may want to update the weights in some special ways, such as dropout. If gradient descent is used to optimize this problem, the gradient will be passed as updates
private:
    Matrix data; // matrix is just an abstract type. It could be an object list of vector etc.
    Vector w;
};
```

Then, to train the model, first initialize the Model object, and then call something like `some_optimization_algo(model, regularization_func_grad / prox_func, some, parameters, for, that, algo)`.

In this way, each type of model corresponding to one `Model` class. However, this is not rigorous enough. In fact, for the same model, the cost function can be different, e.g. linear regression vs least absolute deviations (for both of them the model is f(x) = w' x + b).

### 3.1 Setting at runtime
An optimization algorithm is used in this way:
`xxx_opt_algo(cost_func, cost_grad_func, cost_partial_grad_func, regularization_func_grad / prox_func, some, parameters, for, the, algo);`
or
```C++
XXXOptAlgo xxx_opt_algo;
xxx_opt_algo.set_cost_func(cost_func);
xxx_opt_algo.set_cost_grad_func(cost_grad_func);
...
```

### 3.2 Using Subclass
Each optimization function is an abstract class like that:

```C++
class XXXOptAlgo {
public:
    virtual double cost_func() = 0;
    virtual Vector cost_func_grad() = 0;
    virtual Vector cost_func_grad(int i) = 0; // for something like SGD
    virtual double regularization_func() = 0; // seems that it is not needed in most cases
    virtual double regularization_func_grad() = 0;
    virtual double prox_func() = 0;
    virtual void update_w(Vector updates) = 0; // the problem may want to update the weights in some special ways, such as dropout. If gradient descent is used to optimize this problem, the gradient will be passed as updates
    void train(some, parameters);
};
```

For each optimization problem, the user inherit this class and implement the virtual functions. Then, construct the object and call `train(some, parameters);` In fact this is similar to the first one, but this is a bit clumsy.

**Comment by Kelvin-Ng** [comment]
[Performance issue](https://github.com/husky-team/husky/wiki/TODO:-Format-of-Machine-Learning-API#performance-issue) is added.

**Comment by ddmbr** [comment]
Somehow I feel this design is limited to gradient-based method only? I think all first-order gradient-based methods fit into this class. Is this your intention?

If so, we can actually create different type traits for different classes of problems.

**Comment by Kelvin-Ng** [comment]
I said that this is for example only.

For genetic algorithm, for example again, the `Problem` class should contain public methods of `mutation()`, `crossover()`, etc. I was only going to explain what *kind* of methods should be included in the classes, not exactly what methods should be included.

If we implement genetic algorithm later, we need to add `mutation()`, `crossover()`, etc, to the `Problem` classes existing in the husky lib so that users can choose to use gradient-based methods or genetic algorithm.

**Comment by TatianaJin** [comment]
Your idea is really inspiring. I am very interested in how to generalize the design to non-gradient-based methods. For example, to me the optimization problem for k-means like algorithms has a quite specific logic to find the optimum. My current work (yes they are messy especially the svm one) represents the optimization problem by the gradient function in models and let optimizers handle which samples to use and the update sequence. And a model assigns its problem to an optimizer. I do not know how to extend this to other optimization methods... Could you explain more on `mutation()` and `crossover()` (forgive my ignorance)?

And I think the optimization problems are only part of a model? For some machine learning models the inference parts are the same while the optimization part can be alternated. And I considered them as the same model with different formulations.

Also, if considering neural network (I am not sure whether you think this belongs to the same topic), a model may contain many 'submodels' if each neuron is considered a model according to its activation function, and each activation function corresponds to a type of optimization problem. It appears to me the whole machine learning algorithm takes the form: a model assigns the optimization problems to different types of optimizers and uses the solution to do inference.

**Comment by lijinf2** [comment]
@Kelvin-Ng I agree with you. Yuzhen and me once developed a general framework for genetic algorithms for AI course, and it works.  For each class of machine learning algorithms, I believe there should be a general, efficient and user-friendly framework, which can be built upon Husky. But there is no such framework that is general, efficient and user-friendly to all machine learning algorithms. I think any of the three methods you mentioned is reasonable and no one is always better than the other two. Why not we first determine what class of ml algorithms we want to support, and then consider which of the three method is the best.

**Comment by Kelvin-Ng** [comment]
@TatianaJin 

Your last sentence precisely and concisely summarize what machine learning is basically doing.

What you say is very correct. Sometimes one machine learning model can be optimized in different ways by modelling it as different optimization problems (e.g. L1 vs L2 regularized linear regression), or by using different optimization algorithms on the same model and same optimization problem (e.g. L2 regularized linear regression using FGD, SGD, SVRG or genetic algorithm). The best design should be able to separate all of these so that models, optimization problems and optimization algorithms can be combined in different ways flexibly.

**Comment by Kelvin-Ng** [comment]
The wiki page is updated with a new section. Thanks @TatianaJin for inspiring me.
